### PR TITLE
MAINT: add muon-fmt pre-commit hook

### PR DESCRIPTION
**Pull request by FFY00**
Similarly to #347, it'd also be good to have an autoformatter for the meson files, to ensure we have a consistent style.

The only worry is that muon currently only supports Meson 0.64.0, which will limit us to that syntax, but I don't think we need any fancy features anyway, so I am not worried about that.

### CODE REVIEW

**Comment by FFY00** [comment]
It seems pre-commit.ci doesn't support docker-based actions.

File: docs/examples/spam/meson.build
Comment (dnicolodi) [inline docs/examples/spam/meson.build:24]: I don't like how muon-fmt likes to format the code. More about this later, but can it at least be made to be consistent with itself? For example, here it prefers to have a list of one element with the delimiters inline and additional arguments to follow...

**Review by dnicolodi — CHANGES_REQUESTED** [review]
I really like the effort to make the style of the meson source code more uniform and more polished. Thank you. However, it seems that muon-fmt chooses a fairly poor style for some code. I don't think it is worth to have to manually format these and have markers to tell muon-fmt to leave them alone, it is clutter in the source code for very little gain, also because these files do not really change often, if ta all.

Unless there is a way to configure muon-fmt to do the right thing in these instances (I haven't check, maybe there is). I would commit the result of running muon-fmt, except the instances where it make things worst, and I would not add the pre-commit hook.

While we are at it, we may also consider turning instances of
```meson
py_mod = import('python')
py = py_mod.find_installation()
```
into
```
py = import('python').find_installation()
```
I like the latter much more.

File: docs/examples/spam/meson.build
Comment (rgommers) [inline docs/examples/spam/meson.build:7]: Meson's recommended style (what it uses in all its docs is 2 spaces), not 4. So the formatter seems to be in the wrong here.

Comment (FFY00) [reply docs/examples/spam/meson.build:7]: Being is used in the docs and being recommended are two very different things. I could not find anything relating to 2 vs 4 spaces in the actual [Meson style recommendations page](https://mesonbuild.com/Style-guide.html), so I would not say this style is "wrong".

At the end of the day, it doesn't really matter anyway.

**Comment by FFY00** [comment]
All the objections here are purely subjective, and I agree with many of them, but AFAIK there's no alternative Meson formatter, so we don't really have any alternative anyway.

The question here is if your dislike for the style, which, again, is purely subjective, is enough to justify the not adapting the auto-formatter at all.

My opinion is that everyone has opinions about code-style, you will never make everyone happy, and this is something that at the end of the day is completely useless to progressing the project anyway.

As such, I am +1 on adopting muon fmt, but definitely happy to reconsider moving to a different formatter if one shows up.

I do not wish to argue any further about this, it will not be productive, so it is now up to the rest of the maintainers to make the final call.

File: docs/examples/spam/meson.build
Comment (dnicolodi) [reply docs/examples/spam/meson.build:7]: The indentation is configurable.

**Comment by dnicolodi** [comment]
I don't find auto formatters useful and I think imposing strict stylistic rules is more a nuisance than anything else. Enforcing a code style that we don't like just for the sake of having a tool enforce it does not make much sense. I'm -1 on enforcing the coding style. I'm +1 on making sure that the `meson.build` are formatted with (reasonable) consistency. Most of the proposed changes here are improvements, with the exceptions highlighted above.

**Comment by FFY00** [comment]
I increased the line length to 127, just like we do with Python code, and played with trailing comas to force calls to be expanded, or the opposite, and I think it now looks significantly better.

@dnicolodi, what do you think?

File: tests/packages/configure-data/meson.build
Comment (FFY00) [reply tests/packages/configure-data/meson.build:17]: This was forced because we had a trailing comma there. The trick is to not have a trailing comma at all, so that it inlines the dictionary. black works the same way.

I think the reasoning is that when you are gonna add another argument to the function, it will have to change the style significantly by moving the dictionary brackets, so in these cases you simply remove the trailing comma to allow it to  shrink the dictionary definition.

**Comment by FFY00** [comment]
> While we are at it, we may also consider turning instances of
> 
> ```meson
> py_mod = import('python')
> py = py_mod.find_installation()
> ```
> 
> into
> 
> ```
> py = import('python').find_installation()
> ```
> 
> I like the latter much more.

I agree. If we don't need to acess the Python module, that is better. Can you submit a PR after we get this in?

**Comment by dnicolodi** [comment]
This resolves some of the issues I pointed out, but not all. Also, the new one-line style of the `project()` call is worst than the one with the languages on separate lines. Still, pre-commit-hook fails to run the formatter, so there does not seem to be a way to run the formatter in CI. What are you proposing to merge exactly?

**Comment by FFY00** [comment]
That is because the pre-commit action is Docker-based, which pre-commit.ci doesn't seem to support. I'll fix that later, after the new meson-python release.

File: docs/examples/spam/meson.build
Comment (rgommers) [reply docs/examples/spam/meson.build]: I meant it's used by Meson everywhere, and also by us and other initial adopters like SciPy. So "felt" like a recommendation.

> At the end of the day, it doesn't really matter anyway.

Agreed. I'll resolve this comment - doesn't really matter.

**Comment by rgommers** [comment]
I'm also not really a fan of adding a new auto-formatter, it's just one more hurdle to trip over when submitting a PR. We haven't had any real issues with style in `meson.build` files.

Here is what I propose to merge/close this PR: @dnicolodi to update the style changes he liked / agreed with, merging those, and dropping the rest.

**Comment by dnicolodi** [comment]
I submitted #446 to apply uniform coding style to the `meson.build` files.

I applied some manual fixes before running `muon fmt` with this configuration:
```ini
max_line_len = 127
indent_by = '    '
```

As there is not consensus about enforcing the coding style in CI, I'm closing this PR.

**Pull request closed without merging**
